worker/storage/provisioner.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package storage
+
+import (
+	"path/filepath"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"launchpad.net/tomb"
+
+	"github.com/juju/juju/agent"
+	"github.com/juju/juju/api/environment"
+	"github.com/juju/juju/api/provisioner"
+	"github.com/juju/juju/environs/config"
+	corestorage "github.com/juju/juju/storage"
+	"github.com/juju/juju/storage/provider"
+	"github.com/juju/juju/worker"
+)
+
+var logger = loggo.GetLogger("juju.storage.provisioner")
+
+var _ Provisioner = (*storageProvisioner)(nil)
+
+// TODO - add tests
+
+// Provisioner represents a running storage provisioner worker.
+type Provisioner interface {
+	worker.Worker
+
+	// Create makes the specified volumes using a storage provider of a given type.
+	Create(*config.Config, *corestorage.Config, corestorage.ProviderType, []corestorage.VolumeParams) ([]corestorage.BlockDevice, error)
+}
+
+type storageProvisioner struct {
+	api         *environment.Facade
+	machine     *provisioner.Machine
+	agentConfig agent.Config
+	tomb        tomb.Tomb
+}
+
+// Err returns the reason why the provisioner has stopped or tomb.ErrStillAlive
+// when it is still alive.
+func (p *storageProvisioner) Err() (reason error) {
+	return p.tomb.Err()
+}
+
+// Kill implements worker.Worker.Kill.
+func (p *storageProvisioner) Kill() {
+	p.tomb.Kill(nil)
+}
+
+// Wait implements worker.Worker.Wait.
+func (p *storageProvisioner) Wait() error {
+	return p.tomb.Wait()
+}
+
+// Stop stops the provisioner and returns any error encountered while
+// provisioning.
+func (p *storageProvisioner) Stop() error {
+	p.tomb.Kill(nil)
+	return p.tomb.Wait()
+}
+
+// NewStorageProvisioner creates a new storage provisioner.
+func NewStorageProvisioner(api *environment.Facade, machine *provisioner.Machine, agentConfig agent.Config) Provisioner {
+	p := &storageProvisioner{
+		api:         api,
+		machine:     machine,
+		agentConfig: agentConfig,
+	}
+	go func() {
+		defer p.tomb.Done()
+		p.tomb.Kill(p.loop())
+	}()
+	return p
+}
+
+func (p *storageProvisioner) loop() error {
+	// TODO - this is a hack.
+	// Provision initial storage for loop devices.
```

Comment:
In case there's any doubt, this is a *major* hack. We shouldn't be using the machine provisioner API, and we shouldn't be using the machine provisioning info. Instead, we should be watching all unprovisioned storage instances.